### DISCUSSION TITLE
Python pulp error when exporting a problem to json file.

### DISCUSSION THREAD

**Discussion by moheikal79** [coin-or/pulp#419] [Q&A]
Hi,
Please for advice, I am facing a problem when exporting a python pulp problem to json file "-75 is not JSON serializable".

![image](https://user-images.githubusercontent.com/50020019/112725660-70df1a80-8f2a-11eb-9a18-41b9fa15cfaa.png)

variables= pulp.LpVariable.dicts("x", (item for item in df.index), lowBound=0,  cat='Integer')

model = pulp.LpProblem("Nova_2_levels_maximize", LpMaximize)

model += pulp.lpSum([variables[item] * df.loc[item, 'Profitability'] for item in df.index])

max_weight = 500000
model += pulp.lpSum([variables[item] * df.loc[item, 'Weight'] for item in df.index]) <= max_weight

max_size = 100000
model += pulp.lpSum([variables[item] * df.loc[item, 'Size'] for item in df.index]) <= max_size

min_sales = 4000
model += pulp.lpSum([variables[item] * df.loc[item, 'Sales price'] for item in df.index]) >= min_sales

min_qty = df.index
for nq in min_qty:
    model += variables[nq] >= df.loc[nq, 'Min Qty sales']
    
max_qty = df.index
for xq in max_qty:
    model += variables[xq] <= df.loc[xq, 'Max Qty sales']

model.solve()

model.to_json(r"D:\PL\test.json")

**Comment by pchtsp** [comment]
That's interesting. It's possible there's an error in export function (or your model has something we did not expect). Do you have a reproducible example we can test to find the source of the problem?

**Reply by moheikal79** [reply]
Thanks for your reply, I am attaching the excel version of the data use and the code as well.

Regards 

[Nova LP.xlsx](https://github.com/coin-or/pulp/files/6222595/Nova.LP.xlsx)
[pulp+json.txt](https://github.com/coin-or/pulp/files/6222597/pulp%2Bjson.txt)

**Reply by pchtsp** [reply]
Hello again,

I have checked your code. Apparently, the python json module does not like some of the pandas / numpy data types (floats, int64, etc.).

You have to choices: 

**option1**: type casting, e.g.:

```
#(...)
for xq in max_qty:
    model += variables[xq] <= int(df.loc[xq, 'Max Qty sales'])
# you need to change it every time you use a pandas data.frame, using `float` or `int` at each time
#(...)
```

**option2**: pass a custom encoder to the model.toJson function as explained [here](https://stackoverflow.com/questions/50916422/python-typeerror-object-of-type-int64-is-not-json-serializable/50916741): 

```
import numpy as np
#(...)
class NpEncoder(json.JSONEncoder):
    def default(self, obj):
        if isinstance(obj, np.integer):
            return int(obj)
        elif isinstance(obj, np.floating):
            return float(obj)
        elif isinstance(obj, np.ndarray):
            return obj.tolist()
        else:
            return super(NpEncoder, self).default(obj)

model.to_json(r"D:\PL\test.json", cls=NpEncoder)
```

Given that many people actually use `pandas` to formulate problems for pulp, I think I will update the docs with this information.

**Reply by moheikal79** [reply]
Thanks for reply, I have used option 1 and added int() to all constraints and the step of adding variables to the model, and it works fine. It will be a great step to include examples of using pandas along with pulp as they are used together in production. I am attaching the code after resolving the issue. 


[pulp+json.txt](https://github.com/coin-or/pulp/files/6230785/pulp%2Bjson.txt)


Regards

**Comment by pchtsp** [comment]
In fact, I do not like using pandas with pulp. Nevertheless, I know many
people use it since it's easy to pass from tables to model.

I think we had envisioned a "best practices" page somewhere in the docs. We
should one day put something about that.

F.
